p4est baseline (#780)
Ferrite-FEM/Ferrite.jl: docs/src/topics/amr.md

Reviewer: This one is missing an id and I think the heat tutorial should point here for "further information" in an attempt to route the user here (and not into the devdocs).

# Adaptive Mesh Refinement

Adaptive mesh refinement (AMR) is a computational technique used in finite element analysis to enhance the accuracy and efficiency of simulations.
It involves dynamically adjusting the mesh resolution based on some criteria.
By refining the mesh in regions where the solution exhibits features of interest, AMR ensures that computational resources are concentrated where they are most needed, leading to more accurate results without a proportional increase in computational cost.
This refinement can be achieved in different fashions by either adjusting the mesh size (h-adaptivity), the polynomial order of the Ansatz functions (p-adaptivity) or the nodal positions (r-adaptivity).

> In Ferrite.jl, adaptivity is achieved through a p4est type of implementation which covers h-adaptivity.

Reviewer: ```suggestion
Ferrite.jl supports efficient h-adaptivity through a p4est type of implementation.
```

Note that r-adaptivity was always possible if someone is brave enough to implement the error indicator and respective node update functions.

This approach is designed to handle unstructured hexahedral (in 3D) and quadrilateral (in 2D) meshes.

> A further restriction of the p4est type of implementation is isotropic refinement, meaning that elements are subdivided into smaller elements of the same shape.

Reviewer: I get the idea, but please ask Claude or so to clarify this. This statement is probably more confusing than it helps. If we talk about shape as in RefQuadrilateral, then anisotropic refinement also does gives you the same shape. If we talk about same shape as in they are 1:1 copies of each other, then this is just true for cubes and squares.

As soon as higher polynomial degrees are involved, things become more involved.
In Ferrite, a conformity constraint can be constructed with the ConstraintHandler when using a DofHandler which has been constructed with a grid passed from `Ferrite.creategrid(adaptive_grid::ForestBWG)`.
This conformity constraint ensures that each hanging node is constrained appropriately.

```julia
ch = ConstraintHandler(dh)
add!(ch, Ferrite.ConformityConstraint(:u))
```

## Balancing

Hanging nodes can depend in a nested fashion on other hanging nodes.
This case complicates the `ConformityConstraint` massively and is therefore often prohibited by the act of balancing.
Mesh balancing refers to the process of ensuring a smooth and gradual transition in element sizes across the computational mesh.
This is especially important in adaptive mesh refinement (AMR), where different regions of the mesh may undergo varying levels of refinement based on the solution's characteristics.
The goal of mesh balancing is to prevent isolated regions of excessive refinement, which can lead to inefficiencies in terms of constructing conformity constraints and numerical instability.

> A famous balancing approach is the 2:1 balance, where it is ensured that a hanging node only depends on non-hanging nodes.

Reviewer: I do not think this is accurate. Hanging nodes always depend on non-hanging nodes for nested refinements. In the worst case in a cyclic order, thus tanking the solvability of the linear system.

If I understand everything correctly, then the 2:1 balance described in p4est guarantees that we only need to constraint nodes on faces in 3D to get the correct global continuity. I.e. we do not need to search in addition to the faces which edge neighbor pairs are non-conforming. Think about the first case below in 3D, where the the center node (looking into the screen as the third dimension) hanging into the element on top right is becoming indicrectly constrained now, such that we obtain a nested constraint. If you take a look at the second ascii art below, this does cannot happen anymore. Does this make sense?

```
x-----------x-----------x               x-----------x-----------x
|           |           |               |     |     |     |     |
|           |           |               |     |     |     |     |
|           |           |               |     |     |     |     |
|           |           |               |-----x-----|-----x-----|
|           |           |               |     |     |     |     |
|           |           |               |     |     |     |     |
|           |           |   balancing   |     |     |     |     |
x-----x--x--x-----------x   -------->   x-----x--x--x-----------x
|     |  |  |           |               |     |  |  |     |     |
|     x--x--x           |               |     x--x--x     |     |
|     |  |  |           |               |     |  |  |     |     |
x-----x--x--x           |               x-----x--x--x-----x-----|
|     |     |           |               |     |     |     |     |
|     |     |           |               |     |     |     |     |
|     |     |           |               |     |     |     |     |
x-----x-----x-----------x               x-----x-----x-----------x
```

> In Ferrite's p4est implementation, one can call `balanceforest!` to balance the adaptive grid.

Reviewer: This, or am I missing anything?
```suggestion
In Ferrite's p4est implementation, one must call `balanceforest!` to balance the adaptive grid to ensure all algorithms work correctly.
```

```julia
Ferrite.balanceforest!(adaptive_grid)
```

## Error Estimation
Error estimation is a critical component of adaptive mesh refinement (AMR) in finite element analysis.
The primary objective of error estimation is to identify regions of the computational domain where the numerical solution is less accurate and requires further refinement.
Accurate error estimation guides the adaptive refinement process, ensuring that computational resources are concentrated in areas where they will have the most significant impact on improving solution accuracy.

In practice, error estimators evaluate the local error in the finite element solution by comparing it to an approximation of the true solution.
Common techniques include residual-based error estimation, where the residuals of the finite element equations are used to estimate the local error, and recovery-based methods, which involve constructing a higher-order approximation of the solution and assessing the difference between this approximation and the finite element solution.

> By identifying elements with high estimated errors, these methods provide a targeted approach to mesh refinement, enhancing the overall efficiency and accuracy of the simulation.

Reviewer: Maybe we should also quickly talk that there are different marking strategies and the best one is typically problem dependent.

Author: see below
